### PR TITLE
Add support for Phi3V

### LINKED ISSUE #1108: [Feature] Do we have any plan for supporting Phi3V?
### Checklist

- [ ] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [ ] 2. Please use English, otherwise it will be closed.

### Motivation

Do we have any plan for supporting Phi3V?

### Related resources

_No response_

### PR DESCRIPTION
**Pull request by ravi03071991**
PR to add support for Phi3V. 

Fixes #1108

### CODE REVIEW

**Comment by ravi03071991** [comment]
NOTE: Testing is still pending.

**Comment by merrymercy** [comment]
I added you as a co-author in https://github.com/sgl-project/sglang/pull/2500, so your future commits will trigger CI automatically.

**Comment by merrymercy** [comment]
How did you test the correctness of this model locally? Did you compare the logits against HF implements, similar to this one? https://github.com/sgl-project/sglang/pull/2365#discussion_r1872868614

**Comment by ravi03071991** [comment]
> How did you test the correctness of this model locally? Did you compare the logits against HF implements, similar to this one? [#2365 (comment)](https://github.com/sgl-project/sglang/pull/2365#discussion_r1872868614)

Thanks for the pointer @merrymercy. I'll test it out.

Currently, it seems that Phi3VConfig and Phi3VModel are not part of the transformers library. Instead, they reside in the model files. As a result, I need to add them to the transformers library first to make this PR functional.

In other words, the following imports currently throw an error:

```python
from transformers import Phi3VConfig, Phi3Vmodel
````

**Comment by merrymercy** [comment]
@ravi03071991 Can you fix the error in CI? https://github.com/sgl-project/sglang/actions/runs/12452217969/job/34760948018?pr=2383#step:4:984 Please make sure you can run it locally.

**Comment by ravi03071991** [comment]
@merrymercy yeah sure. I am still working on it.

**Comment by ravi03071991** [comment]
Updates:

1. Added an image processing step prior to image embedding.
2. Implemented image embedding.
3. Integrated Phi3VForCausalLM functionalities:
          - Loading weights
          - Embedding text and images
          - Combining text and image embeddings and passing them to the LLM.

TODO:

The logic for combining text and image embeddings remains unclear. I have imported the logic from [Qwen2 VL](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/models/qwen2_vl.py), but it differs from the approach suggested in the [Hugging Face Phi3V code base](https://huggingface.co/microsoft/Phi-3-vision-128k-instruct/blob/c45209e90a4c4f7d16b2e9d48503c7f3e83623ed/image_embedding_phi3_v.py#L214). Specifically, I am struggling to understand the rationale behind combining the embeddings using image offset and padding.

can someone help me here?

**Comment by zhyncs** [comment]
> can someone help me here?

@yizhang2077 may help take a look. Thanks.

**Comment by yizhang2077** [comment]
> can someone help me here?

In function **pad_input_ids**, it do padding for original input_ids with image tokens and add record image offsets here (where a image embedding start from). I think you can do some modification here, and in **forward** you can replace embedding by using image embedding and image offsets

**Comment by ravi03071991** [comment]
> > can someone help me here?
> 
> In function **pad_input_ids**, it do padding for original input_ids with image tokens and add record image offsets here (where a image embedding start from). I think you can do some modification here, and in **forward** you can replace embedding by using image embedding and image offsets

Thanks @yizhang2077. The `pad_input_ids` seems to be defined in [qwen2_vl](https://github.com/sgl-project/sglang/blob/b5fb4ef58a6bbe6c105d533b69e8e8bc2bf4fc3c/python/sglang/srt/models/qwen2_vl.py#L488) but it does not seem to be used?

**Comment by yizhang2077** [comment]
> Thanks @yizhang2077. The `pad_input_ids` seems to be defined in [qwen2_vl](https://github.com/sgl-project/sglang/blob/b5fb4ef58a6bbe6c105d533b69e8e8bc2bf4fc3c/python/sglang/srt/models/qwen2_vl.py#L488) but it does not seem to be used?

It is used in `pad_input_ids_func` [here](https://github.com/sgl-project/sglang/blob/b5fb4ef58a6bbe6c105d533b69e8e8bc2bf4fc3c/python/sglang/srt/managers/scheduler.py#L586)

**Comment by ravi03071991** [comment]
> It is used in `pad_input_ids_func` [here](https://github.com/sgl-project/sglang/blob/b5fb4ef58a6bbe6c105d533b69e8e8bc2bf4fc3c/python/sglang/srt/managers/scheduler.py#L586)

Oh, I see. I missed that. The padding function seems different for qwen2_vl and llava. Is it specific to the model, and should it be checked on HF?

**Comment by ravi03071991** [comment]
Looks like the model provider has some padding logic [here](https://huggingface.co/microsoft/Phi-3-vision-128k-instruct/blob/c45209e90a4c4f7d16b2e9d48503c7f3e83623ed/processing_phi3_v.py#L176).

**Comment by yizhang2077** [comment]
> Oh, I see. I missed that. The padding function seems different for qwen2_vl and llava. Is it specific to the model, and should it be checked on HF?

I think it is specific to the model since how to do padding depends on model implementation

**Comment by ravi03071991** [comment]
@yizhang2077 / @zhyncs  I think I kinda stuck here:

1. pad_input_ids requires num_img_tokens.
2. The current implementation computes num_img_tokens based on the image processor, which is called inside the forward method.

I’m a bit confused about how to proceed from here. Could you help me here?

**Comment by ravi03071991** [comment]
> @yizhang2077 / @zhyncs I think I kinda stuck here:
> 
> 1. pad_input_ids requires num_img_tokens.
> 2. The current implementation computes num_img_tokens based on the image processor, which is called inside the forward method.
> 
> I’m a bit confused about how to proceed from here. Could you help me here?

Okay. I figured out the best way to solve this is by moving image_processing step into `pad_input_ids` and store the necessary information in `image_inputs` so that they can be used later in forward.
